AppDashboard/lib/datastore_viewer.py

```diff
@@ -0,0 +1,637 @@
+                                order, start):
+    """ Fetches template variables for datastore viewer page for a given kind.
+
+    Args:
+      ds_access: A DatastoreDistributed client.
+      request_uri: A string specifying the current request location.
+      kind: A string specifying the entity kind.
+      namespace: A string specifying the datastore namespace.
+      order: A string containing the name of the property to sorted the results
+        by. A "-" prefix indicates descending order e.g. "-age".
+      start: The number of initial entities to skip in the result set.
+    Returns:
+      A tuple containing headers, entities, and total entity count.
+    """
+    entities, total_entities = _get_entities(
+      ds_access, kind, namespace, order, start, cls.NUM_ENTITIES_PER_PAGE)
+
+    property_name_to_value = _property_name_to_value(entities)
+
+    headers = [{'name': property_name}
+               for property_name in sorted(property_name_to_value)]
+
+    template_entities = []
+    for entity in entities:
+      attributes = []
+      for property_name in sorted(property_name_to_value):
+        if property_name in entity:
+          raw_value = entity[property_name]
+          data_type = DataType.get(raw_value)
+          value = data_type.format(raw_value)
+          short_value = data_type.short_format(raw_value)
+        else:
+          value = ''
+          short_value = ''
+        attributes.append({'name': property_name,
+                           'value': value,
+                           'short_value': short_value,
+                          })
+      edit_uri = '/datastore_viewer/%s/edit/%s?next=%s' % (
+          ds_access.project_id, entity.key(), urllib.quote(request_uri))
+      template_entities.append(
+          {'attributes': attributes,
+           'edit_uri': edit_uri,
+           'key': entity.key(),
+           'key_id': entity.key().id(),
+           'key_name': entity.key().name(),
+           'shortened_key': str(entity.key())[:8] + '...',
+           'write_ops': cls._get_write_ops(ds_access, entity)})
+    return headers, template_entities, total_entities
+
+  @classmethod
+  def _get_indexes(cls, ds_access):
+    """ Retrieves a list of composite indexes for a project.
+
+    Args:
+      ds_access: A DatastoreDistributed client.
+    Returns:
+      A list of datastore.Index objects.
+    """
+    request = api_base_pb.StringProto()
+    request.set_value(ds_access.project_id)
+    response = datastore_pb.CompositeIndices()
+    ds_access._Dynamic_GetIndices(request, response)
+    indexes = []
+    for index in response.index_list():
+      props = [(prop.name(), prop.direction())
+               for prop in index.definition().property_list()]
+      new_index = datastore.Index(index.id(), index.definition().entity_type(),
+                                  index.definition().ancestor(), props)
+      indexes.append((new_index, None))
```

Comment:
why is this a tuple if the second value is always going to be `None`?
  
Also technically doesn't this mean the docstring should say returns a list of tuples?

Reply:
The second item in the tuple was indicating index state sometimes in the SDK. I just used the same signature to avoid modifying the method that used it.